src/local-ssh/ipc/extension.ts

```diff
@@ -0,0 +1,141 @@
+            this.usePublicApi ? this.sessionService.getAPI().getOwnerToken(workspaceId) : svc.server.getOwnerToken(workspaceId),
+        ]), this.logService);
+
+        const phase = this.usePublicApi ? phaseMap[(workspace as Workspace).status?.instance?.status?.phase ?? WorkspaceInstanceStatus_Phase.UNSPECIFIED] : (workspace as WorkspaceInfo).latestInstance?.status.phase;
+        if (phase !== 'running') {
+            const flow: UserFlowTelemetry = { workspaceId, gitpodHost, userId: this.sessionService.getUserId(), flow: 'extension_ipc' };
+            // TODO: show notification only once
+            showWsNotRunningDialog(workspaceId, gitpodHost, flow, this.notificationService, this.logService);
+            throw new ServerError(Status.UNAVAILABLE, 'workspace is not running, current phase: ' + phase);
+        }
+
+        const ideUrl = this.usePublicApi ? (workspace as Workspace).status?.instance?.status?.url : (workspace as WorkspaceInfo).latestInstance?.ideUrl;
+        if (!ideUrl) {
+            throw new ServerError(Status.DATA_LOSS, 'no ide url found');
+        }
+        const url = new URL(ideUrl);
+        const workspaceHost = url.host.substring(url.host.indexOf('.') + 1);
+        return {
+            workspaceId,
+            workspaceHost,
+            ownerToken,
+        };
+    }
+
+    // TODO(hw): get from experiment
+    public usePublicApi: boolean = false;
+
+    constructor(private logService: ILogService, private sessionService: ISessionService, private hostService: IHostService, private notificationService: INotificationService) { }
+}
+
+export class ExtensionServiceServer extends Disposable {
+    private server: Server;
+
+    private localSSHServiceClient = createClient(LocalSSHServiceDefinition, createChannel(getLocalSSHIPCHandleAddr()));
+    public publicApi: GitpodPublicApi | undefined;
+    private readonly id: string = Math.random().toString(36).slice(2);
+    constructor(
+        private readonly logService: ILogService,
+        private readonly sessionService: ISessionService,
+        private readonly hostService: IHostService,
+        private readonly notificationService: INotificationService,
+    ) {
+        super();
+        this.logService.info('going to start extension ipc service server with id', this.id);
+        const server = createServer();
+        const serviceImpl = new ExtensionServiceImpl(this.logService, this.sessionService, this.hostService, this.notificationService);
+        server.add(ExtensionServiceDefinition, serviceImpl);
+        server.listen(getExtensionIPCHandleAddr(this.id)).then(() => {
+            this.logService.info('extension ipc service server started to listen with id: ' + this.id);
+        }).catch(this.logService.error);
+        this.server = server;
+
+        this.backoffActive();
+
+        setTimeout(() => this.pingLocalSSHService(), 1000);
```

Comment:
should it be it interval?

Reply:
We do if for purpose, to avoid we get two `ping` function running. (inside `pingLocalSSHService`, it will call it self again)